### PR TITLE
Feature/replace moderator entity to user entity with many to many relation to roles entity

### PR DESCRIPTION
**Pull request by khlopkov**

### CODE REVIEW

File: Arcadia.Ask/Storage/DatabaseContext.cs [hunk 1]
```diff
@@ -27,12 +27,29 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
                     PostedAt = DateTimeOffset.Now
                 }
             );
+
+            modelBuilder.Entity<UserRoleEntity>()
```
Comment (santee) [inline Arcadia.Ask/Storage/DatabaseContext.cs]: now, i've added EF migrations support to the application. You can read about it here - https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/
Please add migration for these changes.

File: Arcadia.Ask/Auth/SignInService.cs [hunk 2]
```diff
@@ -33,7 +35,11 @@ public async Task<bool> IsModeratorWithCredentialsExists(string login, string pa
         {
             var hashedPassword = ComputeHashFromString(password);
 
-            return await this.dbCtx.Moderators
```
Comment (sponkrashin) [inline Arcadia.Ask/Auth/SignInService.cs]: As logic is becoming more complex, I'd suggest to move all database related code to separate classes like repositories.

File: Arcadia.Ask/Models/Entities/RoleEntity.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    using System.ComponentModel.DataAnnotations;
+
+    public class RoleEntity
+    {
+        [Key]
+        public Guid RoleId { get; set; }
+
+        public string Name { get; set; }
+
+        public IEnumerable<UserRoleEntity> UserRoles { get; set; }
```
Comment (sponkrashin) [inline Arcadia.Ask/Models/Entities/RoleEntity.cs]: To explicitly load UserRoles property only when it is needed with using `Include` method, it should be `virtual`.
Also, to be able to modify UserRoles later (if ever needed) it should be of `ICollection` type.

File: Arcadia.Ask/Storage/Users/UserRepository.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            this.dbCtx = dbCtx;
+        }
+
+        public async Task<UserEntity> FindUserByLoginAndRole(string login, string role, CancellationToken? token = null)
+        {
+            var found = await this.dbCtx.Users
+                .Include(u => u.UserRoles)
+                .ThenInclude(ur => ur.Role)
+                .Where(u =>
+                    u.UserRoles.Any(ur => ur.Role.Name == RoleNames.Moderator) && u.Login == login).FirstOrDefaultAsync(token ?? CancellationToken.None);
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/Users/UserRepository.cs]: Move `FirstOrDefaultAsync` to the next line, it would look nicer.

Comment (sponkrashin) [reply Arcadia.Ask/Storage/Users/UserRepository.cs]: Also you can replace `Where(...).FirstOrDefaultAsync(token)` with `FirstOrDefaultAsync(..., token)`.

File: Arcadia.Ask/Storage/Users/UserRepository.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+            if (found == null)
+            {
+                return null;
+            }
+
+            this.dbCtx.Entry(found).State = EntityState.Detached;
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/Users/UserRepository.cs]: You don't need this, just use `AsNoTracking` method in your query.

File: Arcadia.Ask/Models/Entities/RoleEntity.cs (hunk 3 above)
Comment (sponkrashin) [reply Arcadia.Ask/Models/Entities/RoleEntity.cs]: You changed `UserEntity` class, but missed this class.

File: Arcadia.Ask/Storage/DatabaseContext.cs (hunk 1 above)
Comment (sponkrashin) [reply Arcadia.Ask/Storage/DatabaseContext.cs]: @khlopkov, you haven't resolved this. When you make changes in the database code-first, you need to add migration to reflect those changes and to be able to apply changes to the database itself.
Now you have changes in code-first part, but database will not know about them.

File: Arcadia.Ask/Storage/Users/IUserRepository.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Arcadia.Ask.Storage.Users
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Arcadia.Ask.Models.Entities;
+
+    public interface IUserRepository
+    {
+        Task<UserEntity> FindUserByLoginAndRole(string login, string role, CancellationToken token = default(CancellationToken));
```
Comment (santee) [inline Arcadia.Ask/Storage/Users/IUserRepository.cs]: what's the purpose of this class and method? If you need authorization, you should use AuthorizationHandlers and policies https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-2.2

Then, you'd have your `SigninService.Login(user, password)` that returns your user with all the roles or null (not bound to any role-specific logic like you have now). Roles then can be set as claims and kept in a cookie

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: https://github.com/SoftwareCountry/arcadia-ask/pull/34
In this branch I changed `SignInService`. Now it has method `GetModeratorByCredentials(...)` which return User with roles and with logins, and then add it to claims.
I didn't understand about Policies and authorization handlers. As I understood, they are needed for `[Authorize]` attribute, so is adding them necessary, if I already check permissions using `[Authorize(Role=roleName)]`

Comment (sponkrashin) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: This method is needed to separate database logic and authorization logic. No matter how you do it, you anyway will need the method to get user from the database.

Comment (santee) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: I'm not sure if you need `IUserRepository`. You have SignIn that may take login and password, why is it not enough?

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: To ensure, that user with such password and login is registered in database

Comment (sponkrashin) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: To not mix database logic and authorization logic. Database logic is very simple - return an entity(-ies) by some conditions. Authorization logic - check that database entity is correct. SRP, remember? ;)

Comment (santee) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: >To ensure, that user with such password and login is registered in database
but this method takes role and login, no password. So it's not used for authentication

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: This is method just queries results by login and role and its needed to split it from business layer, where password is checked. Password checks in method of SignInService, which is responsive for business login.
https://github.com/SoftwareCountry/arcadia-ask/pull/34
In this branch SignInService finds moderator by credentials. It finds user by login and role using repository, then using IPasswordHasher service it checks is password valid and if all actions were successfull, then returns object of User class with collection of roles

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: UserRepository responsible for querying data from database. SignInService responsible for checking credentials and uses this repository

Comment (santee) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: first, UserRepository just complicates things. If you add more methods into it it'll become god-like class full of helper methods.
second, why don't you change logic instead, so it won't be moderator role specific? Still 2 sign in methods, one anonymous, another takes username and password. Call signin, get user from the database with all roles, SignInAsync(prinicipal_with_my_roles).

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: I thought about that, but then decided that sign-in as moderator usecase is specific only for moderator actor (in termenology of UML), so finding specific user encapsulated in own service method. On the other hand sign in can be common method for registered in database users, and controllers or hubs with `[Authorize(Roles="Moderator"]` will be specific use cases for moderators

Comment (santee) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: yes, better make it common method. Now moderator role goes through all the layers, that's not good.

Comment (khlopkov) [reply Arcadia.Ask/Storage/Users/IUserRepository.cs]: I changed signature of UserRepository method, but in this PR it still checks if found user is in role of Moderator. In PR #34 I skip checking roles, and return roles in `IEnumerable` field of `User` class

**Pull request merged**